### PR TITLE
[ASTS] Implementation of SweepBucketPointerTable

### PR DESCRIPTION
**Pull request by mdaudali**
## General
**Before this PR**:
Breaking up https://github.com/palantir/atlasdb/pull/7313 into individual PRs representing each table. They will all be in the same class.

We have various interfaces representing logical tables necessary for tracking bucket state, but no concrete implementation.
**After this PR**:
SweepBucketPointerTable

==COMMIT_MSG==
==COMMIT_MSG==

**Priority**: P2

**Concerns / possible downsides (what feedback would you like?)**:
See https://github.com/palantir/atlasdb/pull/7313#discussion_r1786801324

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
N/A
**What was existing testing like? What have you done to improve it?**:
Added tests
**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
N/A

## Execution
Not wired up.

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No
**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No
**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
Not this one
## Development Process
**Where should we start reviewing?**:
DBA
**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:
N/A
**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 1]
```diff
@@ -81,6 +97,83 @@ public BucketStateAndIdentifier getBucketStateAndIdentifier() {
...
+                                    "No starting bucket found for shard and strategy {}",
+                                    SafeArg.of("shardAndStrategy", shardAndStrategy));
+                            return Bucket.of(shardAndStrategy, 0);
+                        }))
+                .values()
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    // TODO: See if we should centralise the work that was done in Jeremy's PR
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java]: Thoughts on centralising this "bump upwards" logic vs duplicating it just once more here? This is a total of two usages, which isn't that many. pro: more specific log lines and naming, con: need to test again (but it's not that hard to write the tests)

Given it's only two usages, a reasonable option is to deduplicate if there's yet another need for this style of logic.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 2]
```diff
@@ -81,6 +97,83 @@ public BucketStateAndIdentifier getBucketStateAndIdentifier() {
...
+                .collect(Collectors.toMap(
+                        Function.identity(), SweepAssignedBucketStoreKeyPersister.INSTANCE::sweepBucketPointerCell));
+        Map<Cell, Long> results = readCells(shardAndStrategyToCell.values(), bucketIdentifierPersister::tryDeserialize);
+        return KeyedStream.stream(shardAndStrategyToCell)
+                .map(cell -> Optional.ofNullable(results.get(cell)))
+                .map((shardAndStrategy, bucketIdentifier) -> bucketIdentifier
+                        .map(id -> Bucket.of(shardAndStrategy, id))
+                        .orElseGet(() -> {
+                            log.warn(
+                                    "No starting bucket found for shard and strategy {}",
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java]: We should probably say what we're doing (starting from 0), and also that this is expected when starting with autoscaling targeted sweep for the first time? Wouldn't want operators to get too alarmed if they see a pile of these on startup

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java [hunk 3]
```diff
@@ -92,4 +101,65 @@ public void updateStateMachineForBucketAssignerModifiesStateToNewIfOriginalMatch
...
+
+        store.updateStartingBucketForShardAndStrategy(existingBucket);
+        store.updateStartingBucketForShardAndStrategy(newBucket);
+
+        assertThat(store.getStartingBucketsForShards(Set.of(existingBucket.shardAndStrategy())))
+                .containsExactly(existingBucket);
+    }
+
+    @Test
+    public void updateStartingBucketForShardSetsToAtLeastExistingValue() {
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java]: I think this one is more about setting to at least the new value? The shard-and-strategy for both existing and new are the same here

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 4]
```diff
@@ -81,6 +97,83 @@ public BucketStateAndIdentifier getBucketStateAndIdentifier() {
...
+                                    + " need to be re-done.",
+                            SafeArg.of("attemptedNewStartingBucket", newStartingBucket),
+                            SafeArg.of("numAttempts", CAS_ATTEMPT_LIMIT),
+                            e);
+                    throw e;
+                } else {
+                    log.info(
+                            "Failed to read or update starting bucket as part of sweep. Retrying",
+                            SafeArg.of("newStartingBucket", newStartingBucket),
+                            SafeArg.of("attemptNumber", attempt + 1),
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java]: nit: usually for this kind of situation I prefer to have my loop index begin at 1 and proceed to `<= max_attempts`

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java]: Yeah I think 3 is usually my threshold for something getting annoying enough to extract

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 5]
```diff
@@ -18,40 +18,56 @@
...
+            ObjectPersister<BucketStateAndIdentifier> bucketStateAndIdentifierPersister,
+            ObjectPersister<Long> bucketIdentifierPersister) {
         this.keyValueService = keyValueService;
         this.bucketStateAndIdentifierPersister = bucketStateAndIdentifierPersister;
+        this.bucketIdentifierPersister = bucketIdentifierPersister;
     }
 
     static DefaultSweepAssignedBucketStore create(KeyValueService keyValueService) {
         // We explicitly don't want to fail on unknown properties;
         ObjectMapper smileMapper = ObjectMappers.newClientSmileMapper();
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java]: as mentioned in other PR: let's make sure this has a consistent ordering (whether here or elsewhere)

**Review by jeremyk-91 — COMMENTED** [review]
Generally looks good!

**Review by jeremyk-91 — APPROVED** [review]

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java (hunk 4 above)
Comment (mdaudali) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java]: I'll change the original code (https://github.com/palantir/atlasdb/blob/develop/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/progress/DefaultBucketProgressStore.java#L122) to match in another PR!

**Comment by mdaudali** [comment]
👍

**Pull request merged**